src/device_handle/async_api.rs

```diff
@@ -0,0 +1,191 @@
+            // Also, closure will be invoked from whatever thread polls, which
+            // may be different from the current thread. So it must be `Send`.
+            // Also, although many threads at once may poll concurrently, only
+            // one will actually ever execute the transfer at a time, so we do
+            // not need to worry about simultaneous writes to the buffer
+            let closure_as_ptr: *mut F = {
+                let ptr: *const F = &result.closure;
+                ptr as *mut F
+            };
+            // Step 2 of async api
+            ffi::libusb_fill_bulk_transfer(
+                ptr.as_ptr(),
+                device.as_raw(),
+                endpoint,
+                buffer.as_ptr() as *mut u8,
+                buffer.len().try_into().unwrap(),
+                Self::transfer_cb,
+                closure_as_ptr.cast(),
+                timeout,
+            )
+        };
+        result
+    }
+
+    /// Submits a transfer to libusb's event engine.
+    /// # Panics
+    /// A transfer should not be double-submitted! Only re-submit after a submission has
+    /// returned Err, or the callback has gotten an Err.
+    // Step 3 of async API
+    #[allow(unused)]
+    pub fn submit(&mut self) -> Result<(), TransferError> {
+        let errno = unsafe { ffi::libusb_submit_transfer(self.ptr.as_ptr()) };
+        use ffi::constants::*;
+        match errno {
+            0 => Ok(()),
+            LIBUSB_ERROR_BUSY => {
+                panic!("Do not double-submit a transfer!")
+            }
+            LIBUSB_ERROR_NOT_SUPPORTED => Err(TransferError::Other("Unsupported transfer!")),
+            LIBUSB_ERROR_INVALID_PARAM => Err(TransferError::Other("Transfer size too large!")),
+            LIBUSB_ERROR_NO_DEVICE => Err(TransferError::Disconnected),
+            _ => Err(TransferError::Errno("Unable to submit transfer. ", errno)),
+        }
+    }
+
+    // We need to invoke our closure using a c-style function, so we store the closure
+    // inside the custom user data field of the transfer struct, and then call the
+    // user provided closure from there.
+    // Step 4 of async API
+    extern "system" fn transfer_cb(transfer: *mut ffi::libusb_transfer) {
+        // Safety: libusb should never make this null, so this is fine
+        let transfer = unsafe { &mut *transfer };
+
+        // sanity
+        debug_assert_eq!(
+            transfer.transfer_type,
+            ffi::constants::LIBUSB_TRANSFER_TYPE_BULK
+        );
+
+        // sanity
+        debug_assert_eq!(
+            std::mem::size_of::<*mut F>(),
+            std::mem::size_of::<*mut c_void>(),
+        );
+        // Safety: The pointer shouldn't be a fat pointer, and should be valid, so
+        // this should be sound
+        let closure = unsafe {
+            let closure: *mut F = std::mem::transmute(transfer.user_data);
+            &mut *closure
+        };
```

Comment:
>The callback may even still be called with a transfer completion after cancellation if it was in the middle of completing at the time of cancellation.
Are you saying this ordering is possible:

1. Submit transfer
2. Drop
3. Callback informs of cancellation
4. Callback informs of transfer completion

I wouldn't expect 4 to ever come before 3 - are you sure that is possible? If not possible, perhaps I could actually intentionally leak the struct with Box::leak, and then get the box back inside the cancellation callback and properly drop the components?

Reply:
The callback should only get called once for a single transfer (until resubmitted), so you won't get both 3 and 4. What I'm not sure about is if you can get a `LIBUSB_TRANSFER_COMPLETED` callback after calling `libusb_transfer_cancel` if the transfer has already completed but the event loop has not called the callback yet, or whether this would still be `LIBUSB_TRANSFER_CANCELLED` but have nonzero `actual_length`.

I think deferring the drop to the callback could work, though I'd probably use `ManuallyDrop` over `Box::leak` to make it explicit.